File: python/binary-search/binary_search.py
def find_errors(search_list, value):
    if len(search_list) == 0:
        raise ValueError("value not in array")
    elif value < search_list[0] or value > search_list[-1]:
        raise ValueError("value not in array")

def find(search_list, value):
    find_errors(search_list, value)

    low = search_list[0]
    high = search_list[-1]
    middle = search_list[len(search_list) // 2]
    while low <= high:
        if value == middle:
            return search_list.index(middle)
        elif value < middle:
            high = search_list[search_list.index(middle) - 1]
            middle = search_list[(search_list.index(low) + search_list.index(high)) // 2]
        else:
            low = search_list[search_list.index(middle) + 1]
            middle = search_list[(search_list.index(low) + search_list.index(high)) // 2]
    if high < low:
        raise ValueError("value not in array")

File: c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length)
{
    // Initialize pointers to the start and end of the array
    int l = 0;
    int r = length - 1;

    // While we can still search the array
    while (l <= r)
    {
        // Calculate the middle index
        int mid = l + ((r - l) / 2);

        // Check if the value is in the middle array
        if (arr[mid] == value)
        {
            return &arr[mid];
        }

        // Check the left side if value < arr[mid]
        if (value < arr[mid])
        {
            r = mid - 1;
        }

        // Check the right side if value > arr[mid]
        if (value > arr[mid])
        {
            l = mid + 1;
        }
    }

    // If we didn't find the value in the array
    return NULL;
}